type-check/groupBy.test.ts

```diff
@@ -30,9 +30,74 @@ const res4 = pipe(
   groupBy((a) => Promise.resolve(a.name)),
 );
 
+type Res5To8 = {
+  a: {
+    type: "a";
+    value: number;
+  }[];
+  b: {
+    type: "b";
+    value: number;
+  }[];
+  c: {
+    type: "c";
+    value: number;
+  }[];
+};
+
+type Res9To12 = { a: "a"[]; b: "b"[]; c: "c"[] };
+
+const list1 = [
+  { type: "a", value: 1 },
+  { type: "b", value: 2 },
+  { type: "c", value: 3 },
+  { type: "a", value: 4 },
+  { type: "b", value: 5 },
+  { type: "c", value: 6 },
+  { type: "a", value: 7 },
+  { type: "b", value: 8 },
+  { type: "c", value: 9 },
+  { type: "a", value: 10 },
+] as { type: "a" | "b" | "c"; value: number }[];
+const res5 = groupBy((a) => a.type, list1);
+
+const res6 = pipe(
+  list1,
+  groupBy((a) => a.type),
+); // Res
+const res7 = groupBy((a) => a.type, toAsync(list1)); // Promise<Res>
+const res8 = pipe(
+  list1,
+  toAsync,
+  groupBy((a) => a.type),
+); // Promise<Res>
+
+const list2 = ["a", "b", "c", "a", "b", "c"] as Array<"a" | "b" | "c">;
+
+const res9 = groupBy((a) => a, list2); // Res
+const res10 = pipe(
+  list2,
+  groupBy((a) => a),
+); // Res
+
+const res11 = groupBy((a) => a, toAsync(list2)); // Promise<Res>
+const res12 = pipe(
+  list2,
+  toAsync,
+  groupBy((a) => a),
+); // Promise<Res>
+
```

Comment:
I thought of another edge case, please also consider the following case. Perhaps inferring it in the most original form would be the best option.

```ts
const list1 = [
  { type: "a", value: 1, value2: 'a' },
  { type: "b", value: 2, value2: 'b' },
  { type: "c", value: 3, value2: 'c' },
  { type: "a", value: 4, value2: 'a' },
  { type: "b", value: 5, value2: 'b' },
  { type: "c", value: 6, value2: 'c' },
  { type: "a", value: 7, value2: 'a' },
  { type: "b", value: 8, value2: 'b' },
  { type: "c", value: 9, value2: 'c' },
  { type: "a", value: 10, value2: 'a' },
] as { type: "a" | "b" | "c"; value: number; value2: "a" | "b" | "c" }[];

type Res = { type: "a" | "b" | "c"; value: number; value2: "a" | "b" | "c" };

groupBy((a) => a.type, list1); // { "a": Res[]; "b": Res[]; "c": Res[]; } ?

```

Reply:
In my opinion, this case was written incorrectly. Perhaps the intention was to infer that type and value2 are the same. If that is the case, the signal that type and value2 are the same should have been given when declaring the list.

Reply:
```ts
const list3 = [
  { type: "a", value: 1, value2: 'a' },
  { type: "b", value: 2, value2: 'b' },
  { type: "c", value: 3, value2: 'c' },
  { type: "a", value: 4, value2: 'a' },
  { type: "b", value: 5, value2: 'b' },
  { type: "c", value: 6, value2: 'c' },
  { type: "a", value: 7, value2: 'a' },
  { type: "b", value: 8, value2: 'b' },
  { type: "c", value: 9, value2: 'c' },
  { type: "a", value: 10, value2: 'a' },
] as ({ type: "a"; value: number; value2: "a" } | { type: "b"; value: number; value2: "b" } | { type: "c"; value: number; value2: "c" })[];

groupBy((a) => a.type, list3);
```

Reply:
What do you think? @shine1594

Reply:
I'm sorry for the late reply. Our team is currently going through a very difficult and challenging time due to workload. We ask for your understanding. Continuing on from the previous comment, if the user asserts the type explicitly as you have shown in the code, there will be no problem at all.

What I'm concerned about is when the types of the `type` and `value2` fields accidentally match in the example code you provided. In such cases, it is impossible to determine at compile time which field will be used as the key for groupBy.

Therefore, I suggested that inferring the original type as I did above is probably the best approach. As a clear example, we cannot assert that there will be no value like`{ type: "a", value: 10, value2: 'c' }` in the `list`.

So, I think the following test case should be added in conclusion:

```ts
type State = "a" | "b" | "c";
type Item = { type: State; value: number; value2: State };
const list: Item[] = [
  { type: "a", value: 1, value2: "a" },
  { type: "b", value: 2, value2: "b" },
  { type: "c", value: 3, value2: "c" },
  { type: "a", value: 4, value2: "a" },
  { type: "b", value: 5, value2: "b" },
  { type: "c", value: 6, value2: "c" },
  { type: "a", value: 7, value2: "a" },
  { type: "b", value: 8, value2: "b" },
  { type: "c", value: 9, value2: "c" },
  { type: "a", value: 10, value2: "a" },
];
const res = groupBy((a) => a.type, list);
checks([check<typeof res, { a: Item[]; b: Item[]; c: Item[] }, Test.Pass>()]);
```

Reply:
Obviously, when we get a key, we don't know what it is.

We'll fix and mention it later.

Reply:
@shine1594 
I try to do something and don't understand it, so I leave it.

I understand the problem you're raising, but the

but it conflicts with what the issue says.

Reply:
Alright, I'll merge this PR for now and try to modify it based on the code you wrote to satisfy [this test case](https://github.com/marpple/FxTS/pull/199#discussion_r1161289493). I'm really grateful for your contributions so far.